power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp

```diff
@@ -598,17 +598,88 @@ class TapPositionOptimizerImpl<std::tuple<TransformerTypes...>, StateCalculator,
     static constexpr auto transformer_index_of = container_impl::get_cls_pos_v<T, TransformerTypes...>;
     static_assert(((transformer_index_of<TransformerTypes> < sizeof...(TransformerTypes)) && ...));
 
+  public:
+    struct BinarySearch {
+        IntS lower_bound = -12;      // tap_position_left
+        IntS upper_bound = 12;       // tap_position_right
+        IntS current = 0;            // tap_position_middle
+        bool last_down = false;      // last direction
+        bool last_check = false;     // last check
+        bool tap_reverse = false;    // tap direction
+        bool inevitable_run = false; // inevitable run
+
+        BinarySearch() = default;
+        BinarySearch(IntS tap_pos, IntS tap_min, IntS tap_max) { reset_bs(tap_pos, tap_min, tap_max); }
+
+        void reset_bs(IntS tap_pos, IntS tap_min, IntS tap_max) {
+            last_down = false;
+            last_check = false;
+            current = tap_pos;
+            inevitable_run = false;
+            lower_bound = std::min(tap_min, tap_max);
+            upper_bound = std::max(tap_min, tap_max);
+            tap_reverse = tap_max < tap_min;
+        }
+
+        IntS get_bs_tap_left() const { return lower_bound; }
```

Comment:
Why use getters and setters if everything is public?

Reply:
Good question. I have the feeling that the binary search logic would eventually live outside the optimizer, as a generic `tactic` (tentative naming), next to scanline method that is currently present.

Reply:
Tentatively addressed in the latest commit.